baseline-error-prone/src/test/java/com/palantir/baseline/errorprone/IllegalSafeLoggingLambdaTest.java

```diff
@@ -86,6 +86,195 @@ void testNestedAnonymousInLambdaUnsafeExternalData() {
+                            Supplier<@Safe Object> supplier = () -> value + value;
+                          }
+                        }
+                        """)
+                .doTest();
+    }
+
+    @Test
+    public void testFunctionalInterfaceWithVoidReturnType() {
+        helper().addSourceLines(
+                        "Test.java",
+                        // language=Java
+                        """
+                        import com.palantir.logsafe.*;
+                        import java.util.function.*;
+
+                        class Test {
+                          static void f() {
+                            // void-returning lambdas should pass the checks fine
+                            Consumer<@Safe Object> supplier = (value) -> {};
+                          }
+                        }
+                        """)
+                .doTest();
+    }
+
+    @Test
+    public void testFunctionalInterfaceWithMultipleMethods() {
+        helper().addSourceLines(
+                        "Test.java",
+                        // language=Java
+                        """
+                        import com.palantir.logsafe.*;
+                        import java.util.List;
+                        import java.util.function.*;
+
+                        @FunctionalInterface
+                        interface MultiMethod<T> {
+                          T get();
+
+                          default List<T> getMany() {
+                            return List.of(get());
+                          }
+
+                          static <T> MultiMethod<T> of(T value) {
+                            return () -> value;
+                          }
+                        }
+
+                        class Test {
+                          static void f(@Unsafe Object value) {
+                            // BUG: Diagnostic contains: Dangerous return value:
+                            // result is 'UNSAFE' but the lambda expects return 'SAFE'.
+                            MultiMethod<@Safe Object> supplier = () -> value;
+                          }
+                        }
+                        """)
+                .doTest();
+    }
+
+    @Test
+    public void testUnsafeLambdaAsReturnType() {
+        helper().addSourceLines(
+                        "Test.java",
+                        // language=Java
+                        """
+                        import com.palantir.logsafe.*;
+                        import java.util.function.*;
+
+                        class Test {
+                          static Supplier<@Safe Object> f(@Unsafe Object value) {
+                            // BUG: Diagnostic contains: Dangerous return value:
+                            // result is 'UNSAFE' but the lambda expects return 'SAFE'.
+                            return () -> value;
+                          }
+                        }
+                        """)
```

Comment:
Mind adding another test case where we define a custom interface along the liens of `interface MaybeSupplier<T> { Optional<T> get(); }`? I think it will work correctly, but it catches an issue from a previous commit where type-parameters were matched with the return type, rather than using the resolved info

Reply:
Something like this? https://github.com/palantir/gradle-baseline/pull/3048/commits/c97ee70949b869358c40ff7aa0e939af796aa48b

Reply:
Perfect, thanks!